File: lib/claude_bridge.py
"""
claude_bridge.py — Shared wrapper around `claude -p` for all integrations.

Supports two modes:
  - Stateless: ClaudeBridge (original, one-shot per request)
  - Living:    LivingBridge (brain-aware, session-persistent, proactive)

Usage:
    # Stateless (backwards compatible)
    from lib.claude_bridge import ClaudeBridge
    bridge = ClaudeBridge(project_dir="/home/user/my-project")
    response = bridge.ask("Fix the bug in auth.py")

    # Living agent
    from lib.claude_bridge import LivingBridge
    agent = LivingBridge(project_dir="/home/user/my-project")
    response = await agent.ask_as("telegram", "123456", "Fix the bug in auth.py")
"""

import asyncio
import json
import logging
import os
import subprocess
import time
from dataclasses import dataclass, field

logger = logging.getLogger(__name__)


@dataclass
class ClaudeResponse:
    """Structured response from a claude -p invocation."""

    text: str
    exit_code: int
    cost_usd: float = 0.0
    duration_ms: int = 0
    duration_api_ms: int = 0
    num_turns: int = 0
    session_id: str = ""
    is_error: bool = False
    raw: dict = field(default_factory=dict)


class ClaudeBridge:
    """Wrapper around the `claude -p` CLI for programmatic access."""

    _DEFAULT_TIMEOUT = 300

    def __init__(
        self,
        project_dir: str | None = None,
        model: str | None = None,
        allowed_tools: list[str] | None = None,
        max_budget_usd: float | None = None,
        timeout_seconds: int | None = None,
    ):
        self.project_dir = project_dir or os.getcwd()
        self.model = model or os.getenv("CLAUDE_MODEL")
        self.allowed_tools = allowed_tools or os.getenv(
            "CLAUDE_ALLOWED_TOOLS", ""
        ).split(",")
        self.allowed_tools = [t.strip() for t in self.allowed_tools if t.strip()]
        if max_budget_usd is not None:
            self.max_budget_usd = max_budget_usd
        else:
            env_budget = os.getenv("CLAUDE_MAX_BUDGET_USD")
            self.max_budget_usd = float(env_budget) if env_budget else None
        if timeout_seconds is not None:
            self.timeout_seconds = timeout_seconds
        else:
            env_timeout = os.getenv("CLAUDE_TIMEOUT_SECONDS")
            self.timeout_seconds = int(env_timeout) if env_timeout else self._DEFAULT_TIMEOUT

    def _build_command(self, prompt: str, resume_session: str | None = None) -> list[str]:
        """Build the claude CLI command."""
        cmd = ["claude", "-p", prompt, "--output-format", "json"]

        if resume_session:
            cmd.extend(["--resume", resume_session])

        if self.model:
            cmd.extend(["--model", self.model])

        if self.allowed_tools:
            cmd.extend(["--allowedTools", ",".join(self.allowed_tools)])

        if self.max_budget_usd is not None:
            cmd.extend(["--max-budget-usd", str(self.max_budget_usd)])

        return cmd

    def _parse_response(self, stdout: str, exit_code: int) -> ClaudeResponse:
        """Parse JSON output from claude -p."""
        try:
            data = json.loads(stdout)
            return ClaudeResponse(
                text=data.get("result", stdout),
                exit_code=exit_code,
                cost_usd=data.get("cost_usd", 0.0),
                duration_ms=data.get("duration_ms", 0),
                duration_api_ms=data.get("duration_api_ms", 0),
                num_turns=data.get("num_turns", 0),
                session_id=data.get("session_id", ""),
                is_error=data.get("is_error", exit_code != 0),
                raw=data,
            )
        except json.JSONDecodeError:
            return ClaudeResponse(
                text=stdout.strip(),
                exit_code=exit_code,
                is_error=exit_code != 0,
            )

    def ask(self, prompt: str, resume_session: str | None = None) -> ClaudeResponse:
        """Send a prompt to Claude Code and return the response (blocking)."""
        cmd = self._build_command(prompt, resume_session)
        logger.info("Running: %s", " ".join(cmd[:4]) + "...")

        try:
            result = subprocess.run(
                cmd,
                capture_output=True,
                text=True,
                timeout=self.timeout_seconds,
                cwd=self.project_dir,
            )
            response = self._parse_response(result.stdout, result.returncode)

            if result.returncode != 0 and not response.text:
                response.text = result.stderr.strip() or "Unknown error"
                response.is_error = True

            logger.info(
                "Response: %d chars, cost=$%.4f, turns=%d",
                len(response.text),
                response.cost_usd,
                response.num_turns,
            )
            return response

        except subprocess.TimeoutExpired:
            logger.error("Claude timed out after %ds", self.timeout_seconds)
            return ClaudeResponse(
                text=f"Timeout after {self.timeout_seconds}s",
                exit_code=-1,
                is_error=True,
            )
        except FileNotFoundError:
            logger.error("claude CLI not found — is Claude Code installed?")
            return ClaudeResponse(
                text="claude CLI not found. Install: npm install -g @anthropic-ai/claude-code",
                exit_code=-1,
                is_error=True,
            )

    async def ask_async(self, prompt: str, resume_session: str | None = None) -> ClaudeResponse:
        """Async version of ask() for use in bot event loops."""
        cmd = self._build_command(prompt, resume_session)
        logger.info("Running async: %s", " ".join(cmd[:4]) + "...")

        try:
            proc = await asyncio.create_subprocess_exec(
                *cmd,
                stdout=asyncio.subprocess.PIPE,
                stderr=asyncio.subprocess.PIPE,
                cwd=self.project_dir,
            )

            try:
                stdout_bytes, stderr_bytes = await asyncio.wait_for(
                    proc.communicate(), timeout=self.timeout_seconds
                )
            except asyncio.TimeoutError:
                proc.kill()
                await proc.communicate()
                return ClaudeResponse(
                    text=f"Timeout after {self.timeout_seconds}s",
                    exit_code=-1,
                    is_error=True,
                )

            stdout = stdout_bytes.decode()
            response = self._parse_response(stdout, proc.returncode or 0)

            if proc.returncode != 0 and not response.text:
                response.text = stderr_bytes.decode().strip() or "Unknown error"
                response.is_error = True

            return response

        except FileNotFoundError:
            return ClaudeResponse(
                text="claude CLI not found. Install: npm install -g @anthropic-ai/claude-code",
                exit_code=-1,
                is_error=True,
            )


class LivingBridge(ClaudeBridge):
    """Brain-aware, session-persistent bridge for the living agent.

    Extends ClaudeBridge with:
    - Persistent memory (brain.md) injected into every prompt
    - Session continuity per user (resume conversations)
    - Automatic brain updates after each interaction
    - Event logging for history
    """

    def __init__(self, **kwargs):
        super().__init__(**kwargs)
        # Lazy imports to avoid circular deps
        from lib.brain import Brain
        from lib.session_store import SessionStore

        self.brain = Brain()
        self.sessions = SessionStore()

    def _build_living_prompt(self, user_message: str) -> str:
        """Wrap user message with brain context."""
        brain_context = self.brain.get_context_prompt()
        return (
            f"{brain_context}\n\n"
            f"---\n\n"
            f"User message:\n{user_message}\n\n"
            f"---\n\n"
            f"Respond to the user's message. If you learn anything new about "
            f"the user's preferences or if there are important events to "
            f"remember, note them — your brain will be updated after this."
        )

    async def ask_as(
        self, platform: str, user_id: str, message: str
    ) -> ClaudeResponse:
        """Send a message as a specific user with brain context and session resume.

        This is the primary method for the living agent. It:
        1. Loads the brain for context
        2. Looks up the user's session for continuity
        3. Sends the prompt with brain + session
        4. Stores the new session ID
        5. Logs the event to brain history
        """
        # Build prompt with brain context
        prompt = self._build_living_prompt(message)

        # Try to resume user's session
        session_id = self.sessions.get(platform, user_id)

        # Send to Claude
        response = await self.ask_async(prompt, resume_session=session_id)

        # Store new session ID for continuity
        if response.session_id:
            self.sessions.set(platform, user_id, response.session_id)

        # Log event to brain
        short_msg = message[:80] + "..." if len(message) > 80 else message
        platform_label = f"{platform}:{user_id}"
        if response.is_error:
            self.brain.add_event(f"[{platform_label}] Error: {response.text[:100]}")
        else:
            self.brain.add_event(
                f"[{platform_label}] Q: {short_msg} "
                f"(cost=${response.cost_usd:.4f}, turns={response.num_turns})"
            )

        return response

    def ask_as_sync(
        self, platform: str, user_id: str, message: str
    ) -> ClaudeResponse:
        """Sync version of ask_as for non-async contexts."""
        prompt = self._build_living_prompt(message)
        session_id = self.sessions.get(platform, user_id)
        response = self.ask(prompt, resume_session=session_id)

        if response.session_id:
            self.sessions.set(platform, user_id, response.session_id)

        short_msg = message[:80] + "..." if len(message) > 80 else message
        platform_label = f"{platform}:{user_id}"
        if response.is_error:
            self.brain.add_event(f"[{platform_label}] Error: {response.text[:100]}")
        else:
            self.brain.add_event(
                f"[{platform_label}] Q: {short_msg} "
                f"(cost=${response.cost_usd:.4f}, turns={response.num_turns})"
            )

        return response

File: src/lib/claude-bridge.ts
/**
 * claude-bridge.ts — Shared wrapper around `claude -p` for all integrations.
 *
 * Supports two modes:
 *   - Stateless: ClaudeBridge (original, one-shot per request)
 *   - Living:    LivingBridge (brain-aware, session-persistent, proactive)
 */

import { spawn, spawnSync, execFileSync } from "node:child_process";
import fs from "node:fs";
import os from "node:os";
import path from "node:path";
import { Brain } from "./brain.js";
import { SessionStore } from "./session-store.js";

/**
 * Resolve the full path to the `claude` CLI binary.
 * Background processes (systemd, make start) often have a minimal PATH
 * that doesn't include npm global bin directories.
 */
function findClaudeBinary(): string {
  // 1. Check CLAUDE_BIN env var (explicit override)
  if (process.env["CLAUDE_BIN"]) return process.env["CLAUDE_BIN"];

  // 2. Try `which claude` (works if PATH is correct)
  try {
    const result = execFileSync("which", ["claude"], { stdio: "pipe", encoding: "utf-8" });
    const found = result.trim();
    if (found) return found;
  } catch { /* not in PATH */ }

  // 3. Check common installation locations
  const candidates = [
    "/usr/local/bin/claude",
    "/usr/bin/claude",
    path.join(process.env["HOME"] ?? "", ".npm-global", "bin", "claude"),
    path.join(process.env["HOME"] ?? "", ".local", "bin", "claude"),
  ];

  // 4. Scan nvm node versions for claude binary
  const nvmDir = process.env["NVM_DIR"] ?? path.join(process.env["HOME"] ?? "", ".nvm");
  const nvmVersionsDir = path.join(nvmDir, "versions", "node");
  try {
    if (fs.existsSync(nvmVersionsDir)) {
      for (const ver of fs.readdirSync(nvmVersionsDir)) {
        candidates.push(path.join(nvmVersionsDir, ver, "bin", "claude"));
      }
    }
  } catch { /* nvm dir not readable */ }

  for (const candidate of candidates) {
    if (fs.existsSync(candidate)) return candidate;
  }

  // 5. Fallback: just "claude" and hope for the best
  return "claude";
}

const CLAUDE_BIN = findClaudeBinary();
if (CLAUDE_BIN === "claude") {
  console.warn("[claude-bridge] Warning: could not resolve full path to claude CLI. Set CLAUDE_BIN env var if claude is installed in a non-standard location.");
} else {
  console.log(`[claude-bridge] Using claude binary: ${CLAUDE_BIN}`);
}

/** Structured response from a `claude -p` invocation. */
export interface ClaudeResponse {
  text: string;
  exitCode: number;
  costUsd: number;
  durationMs: number;
  durationApiMs: number;
  numTurns: number;
  sessionId: string;
  isError: boolean;
  raw: Record<string, unknown>;
}

/** A single event from `claude -p --output-format stream-json`. */
export interface StreamEvent {
  type: "system" | "assistant" | "user" | "result";
  subtype?: string;
  session_id?: string;
  // assistant message fields
  message?: {
    role?: string;
    content?: Array<{
      type: string;
      tool_use_id?: string;
      name?: string;
      text?: string;
      input?: Record<string, unknown>;
      content?: string;
      is_error?: boolean;
      [key: string]: unknown;
    }>;
    [key: string]: unknown;
  };
  // result fields
  result?: string;
  cost_usd?: number;
  duration_ms?: number;
  duration_api_ms?: number;
  num_turns?: number;
  is_error?: boolean;
  [key: string]: unknown;
}

function makeErrorResponse(text: string): ClaudeResponse {
  return {
    text,
    exitCode: -1,
    costUsd: 0,
    durationMs: 0,
    durationApiMs: 0,
    numTurns: 0,
    sessionId: "",
    isError: true,
    raw: {},
  };
}

export class ClaudeBridge {
  protected static readonly DEFAULT_TIMEOUT = 300;

  readonly projectDir: string;
  readonly workspaceDir: string;
  readonly model: string | undefined;
  readonly allowedTools: string[];
  readonly maxBudgetUsd: number | undefined;
  readonly timeoutSeconds: number;
  readonly dangerouslySkipPermissions: boolean;
  readonly mcpConfig: string | undefined;
  readonly cwdOverride: string | undefined;

  constructor(options: {
    projectDir?: string;
    workspaceDir?: string;
    model?: string;
    allowedTools?: string[];
    maxBudgetUsd?: number;
    timeoutSeconds?: number;
    dangerouslySkipPermissions?: boolean;
    mcpConfig?: string;
    cwdOverride?: string;
  } = {}) {
    this.projectDir = options.projectDir ?? process.cwd();
    this.workspaceDir = options.workspaceDir
      ?? process.env["CLAUDE_WORKSPACE_DIR"]
      ?? path.join(os.homedir(), ".claude-agent", "workspace");
    this.model = options.model ?? process.env["CLAUDE_MODEL"];

    if (options.allowedTools) {
      this.allowedTools = options.allowedTools;
    } else {
      const envTools = process.env["CLAUDE_ALLOWED_TOOLS"] ?? "";
      this.allowedTools = envTools.split(",").map((t) => t.trim()).filter(Boolean);
    }

    if (options.maxBudgetUsd !== undefined) {
      this.maxBudgetUsd = options.maxBudgetUsd;
    } else {
      const envBudget = process.env["CLAUDE_MAX_BUDGET_USD"];
      this.maxBudgetUsd = envBudget ? parseFloat(envBudget) : undefined;
    }

    if (options.timeoutSeconds !== undefined) {
      this.timeoutSeconds = options.timeoutSeconds;
    } else {
      const envTimeout = process.env["CLAUDE_TIMEOUT_SECONDS"];
      this.timeoutSeconds = envTimeout ? parseInt(envTimeout, 10) : ClaudeBridge.DEFAULT_TIMEOUT;
    }

    this.dangerouslySkipPermissions = options.dangerouslySkipPermissions ?? false;
    this.mcpConfig = options.mcpConfig;
    this.cwdOverride = options.cwdOverride;
  }

  /** Build the claude CLI command. */
  buildCommand(prompt: string, resumeSession?: string): string[] {
    const cmd = [CLAUDE_BIN, "-p", prompt, "--output-format", "json"];

    if (resumeSession) {
      cmd.push("--resume", resumeSession);
    }
    if (this.model) {
      cmd.push("--model", this.model);
    }
    if (this.allowedTools.length > 0) {
      cmd.push("--allowedTools", this.allowedTools.join(","));
    }
    if (this.maxBudgetUsd !== undefined) {
      cmd.push("--max-budget-usd", String(this.maxBudgetUsd));
    }
    if (this.dangerouslySkipPermissions) {
      cmd.push("--dangerously-skip-permissions");
    }
    if (this.mcpConfig) {
      cmd.push("--mcp-config", this.mcpConfig);
    }

    return cmd;
  }

  /** Parse JSON output from `claude -p`. */
  parseResponse(stdout: string, exitCode: number): ClaudeResponse {
    try {
      const data = JSON.parse(stdout) as Record<string, unknown>;
      return {
        text: (data["result"] as string) ?? stdout,
        exitCode,
        costUsd: (data["cost_usd"] as number) ?? 0,
        durationMs: (data["duration_ms"] as number) ?? 0,
        durationApiMs: (data["duration_api_ms"] as number) ?? 0,
        numTurns: (data["num_turns"] as number) ?? 0,
        sessionId: (data["session_id"] as string) ?? "",
        isError: (data["is_error"] as boolean) ?? exitCode !== 0,
        raw: data,
      };
    } catch {
      return {
        text: stdout.trim(),
        exitCode,
        costUsd: 0,
        durationMs: 0,
        durationApiMs: 0,
        numTurns: 0,
        sessionId: "",
        isError: exitCode !== 0,
        raw: {},
      };
    }
  }

  /** Send a prompt to Claude Code and return the response (blocking). */
  ask(prompt: string, resumeSession?: string): ClaudeResponse {
    const cmd = this.buildCommand(prompt, resumeSession);
    const [bin, ...args] = cmd;

    const env = { ...process.env };
    delete env["CLAUDECODE"];

    try {
      const result = spawnSync(bin, args, {
        cwd: this.cwdOverride ?? this.workspaceDir,
        encoding: "utf-8",
        timeout: this.timeoutSeconds * 1000,
        maxBuffer: 10 * 1024 * 1024,
        env,
      });

      if (result.error) {
        if ((result.error as NodeJS.ErrnoException).code === "ENOENT") {
          return makeErrorResponse(
            "claude CLI not found. Install: npm install -g @anthropic-ai/claude-code"
          );
        }
        if ((result.error as NodeJS.ErrnoException).code === "ETIMEDOUT") {
          return makeErrorResponse(`Timeout after ${this.timeoutSeconds}s`);
        }
        return makeErrorResponse(result.error.message);
      }

      const response = this.parseResponse(result.stdout ?? "", result.status ?? 1);
      if ((result.status ?? 1) !== 0 && !response.text) {
        response.text = result.stderr?.trim() || "Unknown error";
        response.isError = true;
      }

      return response;
    } catch (err) {
      return makeErrorResponse(String(err));
    }
  }

  /** Async version of ask() for use in bot event loops. */
  async askAsync(prompt: string, resumeSession?: string): Promise<ClaudeResponse> {
    const cmd = this.buildCommand(prompt, resumeSession);
    const [bin, ...args] = cmd;

    const env = { ...process.env };
    delete env["CLAUDECODE"];

    return new Promise<ClaudeResponse>((resolve) => {
      let proc;
      try {
        proc = spawn(bin, args, {
          cwd: this.cwdOverride ?? this.workspaceDir,
          stdio: ["ignore", "pipe", "pipe"],
          env,
        });
      } catch {
        resolve(
          makeErrorResponse(
            "claude CLI not found. Install: npm install -g @anthropic-ai/claude-code"
          )
        );
        return;
      }

      const stdoutChunks: Buffer[] = [];
      const stderrChunks: Buffer[] = [];

      proc.stdout.on("data", (chunk: Buffer) => stdoutChunks.push(chunk));
      proc.stderr.on("data", (chunk: Buffer) => stderrChunks.push(chunk));

      const timer = setTimeout(() => {
        proc.kill("SIGKILL");
      }, this.timeoutSeconds * 1000);

      proc.on("error", (err: NodeJS.ErrnoException) => {
        clearTimeout(timer);
        if (err.code === "ENOENT") {
          resolve(
            makeErrorResponse(
              "claude CLI not found. Install: npm install -g @anthropic-ai/claude-code"
            )
          );
        } else {
          resolve(makeErrorResponse(err.message));
        }
      });

      proc.on("close", (code, signal) => {
        clearTimeout(timer);
        if (signal === "SIGKILL") {
          resolve(makeErrorResponse(`Timeout after ${this.timeoutSeconds}s`));
          return;
        }

        const stdout = Buffer.concat(stdoutChunks).toString("utf-8");
        const stderr = Buffer.concat(stderrChunks).toString("utf-8");
        const exitCode = code ?? 1;
        const response = this.parseResponse(stdout, exitCode);

        if (exitCode !== 0 && !response.text) {
          response.text = stderr.trim() || "Unknown error";
          response.isError = true;
        }

        resolve(response);
      });
    });
  }

  /**
   * Send a prompt with streaming output.
   * Each NDJSON line from stdout is parsed and passed to `onEvent`.
   * Resolves with the final ClaudeResponse when the result event arrives.
   */
  async askStreaming(
    prompt: string,
    onEvent: (event: StreamEvent) => void | Promise<void>,
    resumeSession?: string,
  ): Promise<ClaudeResponse> {
    // Build command with stream-json output format
    const cmd = [CLAUDE_BIN, "-p", prompt, "--output-format", "stream-json", "--verbose"];

    if (resumeSession) {
      cmd.push("--resume", resumeSession);
    }
    if (this.model) {
      cmd.push("--model", this.model);
    }
    if (this.allowedTools.length > 0) {
      cmd.push("--allowedTools", this.allowedTools.join(","));
    }
    if (this.maxBudgetUsd !== undefined) {
      cmd.push("--max-budget-usd", String(this.maxBudgetUsd));
    }
    if (this.dangerouslySkipPermissions) {
      cmd.push("--dangerously-skip-permissions");
    }
    if (this.mcpConfig) {
      cmd.push("--mcp-config", this.mcpConfig);
    }

    const [bin, ...args] = cmd;
    const env = { ...process.env };
    delete env["CLAUDECODE"];

    return new Promise<ClaudeResponse>((resolve, reject) => {
      let proc;
      try {
        proc = spawn(bin, args, {
          cwd: this.cwdOverride ?? this.workspaceDir,
          stdio: ["ignore", "pipe", "pipe"],
          env,
        });
      } catch {
        resolve(makeErrorResponse(
          "claude CLI not found. Install: npm install -g @anthropic-ai/claude-code"
        ));
        return;
      }

      let lastResult: ClaudeResponse | null = null;
      let stderrBuf = "";
      let lineBuffer = "";

      proc.stdout.on("data", (chunk: Buffer) => {
        lineBuffer += chunk.toString("utf-8");
        const lines = lineBuffer.split("\n");
        // Keep the last incomplete line in the buffer
        lineBuffer = lines.pop() ?? "";

        for (const line of lines) {
          const trimmed = line.trim();
          if (!trimmed) continue;
          try {
            const event = JSON.parse(trimmed) as StreamEvent;
            // Fire callback (swallow errors to avoid breaking the stream)
            try { void onEvent(event); } catch { /* ignore callback errors */ }

            if (event.type === "result") {
              lastResult = {
                text: (event.result as string) ?? "",
                exitCode: 0,
                costUsd: (event.cost_usd as number) ?? 0,
                durationMs: (event.duration_ms as number) ?? 0,
                durationApiMs: (event.duration_api_ms as number) ?? 0,
                numTurns: (event.num_turns as number) ?? 0,
                sessionId: (event.session_id as string) ?? "",
                isError: (event.is_error as boolean) ?? false,
                raw: event as unknown as Record<string, unknown>,
              };
            }
          } catch {
            // Not valid JSON — ignore
          }
        }
      });

      proc.stderr.on("data", (chunk: Buffer) => {
        stderrBuf += chunk.toString("utf-8");
      });

      const timer = setTimeout(() => {
        proc.kill("SIGKILL");
      }, this.timeoutSeconds * 1000);

      proc.on("error", (err: NodeJS.ErrnoException) => {
        clearTimeout(timer);
        if (err.code === "ENOENT") {
          resolve(makeErrorResponse(
            "claude CLI not found. Install: npm install -g @anthropic-ai/claude-code"
          ));
        } else {
          resolve(makeErrorResponse(err.message));
        }
      });

      proc.on("close", (code, signal) => {
        clearTimeout(timer);
        if (signal === "SIGKILL") {
          resolve(makeErrorResponse(`Timeout after ${this.timeoutSeconds}s`));
          return;
        }

        if (lastResult) {
          resolve(lastResult);
        } else {
          const exitCode = code ?? 1;
          resolve({
            text: stderrBuf.trim() || "No result received from Claude",
            exitCode,
            costUsd: 0,
            durationMs: 0,
            durationApiMs: 0,
            numTurns: 0,
            sessionId: "",
            isError: true,
            raw: {},
          });
        }
      });
    });
  }
}

/**
 * Brain-aware, session-persistent bridge for the living agent.
 *
 * Extends ClaudeBridge with:
 * - Persistent memory (brain.md) injected into every prompt
 * - Session continuity per user (resume conversations)
 * - Automatic brain updates after each interaction
 * - Event logging for history
 */
export class LivingBridge extends ClaudeBridge {
  readonly brain: Brain;
  readonly sessions: SessionStore;

  constructor(options: {
    projectDir?: string;
    model?: string;
    allowedTools?: string[];
    maxBudgetUsd?: number;
    timeoutSeconds?: number;
  } = {}) {
    super(options);
    this.brain = new Brain();
    this.sessions = new SessionStore();
  }

  /** Wrap user message with brain context. */
  private buildLivingPrompt(userMessage: string): string {
    const brainContext = this.brain.getContextPrompt();
    return (
      `${brainContext}\n\n` +
      `---\n\n` +
      `User message:\n${userMessage}\n\n` +
      `---\n\n` +
      `Respond to the user's message. If you learn anything new about ` +
      `the user's preferences or if there are important events to ` +
      `remember, note them — your brain will be updated after this.`
    );
  }

  /**
   * Send a message as a specific user with brain context and session resume.
   *
   * This is the primary async method for the living agent. It:
   * 1. Loads the brain for context
   * 2. Looks up the user's session for continuity
   * 3. Sends the prompt with brain + session
   * 4. Stores the new session ID
   * 5. Logs the event to brain history
   */
  async askAs(platform: string, userId: string, message: string): Promise<ClaudeResponse> {
    const prompt = this.buildLivingPrompt(message);
    const sessionId = this.sessions.get(platform, userId);
    const response = await this.askAsync(prompt, sessionId ?? undefined);

    if (response.sessionId) {
      this.sessions.set(platform, userId, response.sessionId);
    }

    const shortMsg = message.length > 80 ? message.slice(0, 80) + "..." : message;
    const label = `${platform}:${userId}`;
    if (response.isError) {
      this.brain.addEvent(`[${label}] Error: ${response.text.slice(0, 100)}`);
    } else {
      this.brain.addEvent(
        `[${label}] Q: ${shortMsg} (cost=$${response.costUsd.toFixed(4)}, turns=${response.numTurns})`
      );
    }

    return response;
  }

  /** Sync version of askAs for non-async contexts. */
  askAsSync(platform: string, userId: string, message: string): ClaudeResponse {
    const prompt = this.buildLivingPrompt(message);
    const sessionId = this.sessions.get(platform, userId);
    const response = this.ask(prompt, sessionId ?? undefined);

    if (response.sessionId) {
      this.sessions.set(platform, userId, response.sessionId);
    }

    const shortMsg = message.length > 80 ? message.slice(0, 80) + "..." : message;
    const label = `${platform}:${userId}`;
    if (response.isError) {
      this.brain.addEvent(`[${label}] Error: ${response.text.slice(0, 100)}`);
    } else {
      this.brain.addEvent(
        `[${label}] Q: ${shortMsg} (cost=$${response.costUsd.toFixed(4)}, turns=${response.numTurns})`
      );
    }

    return response;
  }
}
